File: v1/session/session.py
"""
Session 管理器 — 会话生命周期管理与执行入口。

职责：
- 会话的创建、删除、查询、列表管理
- 会话的 prompt() 方法：组装提示词并启动 Agent Loop
- 会话状态和消息历史的持久化
"""

import asyncio
from pathlib import Path
from typing import Optional, List, Dict, Any, Callable

from loguru import logger

from agent.agent import AgentConfig, get_agent_config, AgentRole
from provider.model import ModelClient

from .store import SessionStore, SessionRecord
from .prompt import PromptBuilder
from .loop import AgentLoop


# ============================================================================
#  Session 类
# ============================================================================

class Session:
    """
    会话实例。

    封装一个完整的 Agent 会话，包含：
    - 会话元信息（ID、标题、状态等）
    - 消息历史
    - 提示词组装
    - Agent Loop 执行
    """

    def __init__(
        self,
        record: SessionRecord,
        store: SessionStore,
        model_client: ModelClient,
        agent_config: Optional[AgentConfig] = None,
        session_manager: Optional['SessionManager'] = None,
    ):
        self._record = record
        self._store = store
        self._model_client = model_client
        self._agent_config = agent_config or get_agent_config(record.agent_role)
        self._session_manager = session_manager

    # ── 属性访问 ──────────────────────────────────────────

    @property
    def id(self) -> str:
        """会话 ID"""
        return self._record.session_id

    @property
    def parent_id(self) -> Optional[str]:
        """父会话 ID"""
        return self._record.parent_id

    @property
    def title(self) -> str:
        """会话标题"""
        return self._record.title

    @property
    def status(self) -> str:
        """会话状态"""
        return self._record.status

    @property
    def work_dir(self) -> str:
        """工作目录"""
        return self._record.work_dir

    @property
    def agent_role(self) -> str:
        """Agent 角色"""
        return self._record.agent_role

    @property
    def messages(self) -> List[Dict[str, Any]]:
        """消息历史"""
        return self._record.messages.copy()

    @property
    def record(self) -> SessionRecord:
        """底层记录"""
        return self._record

    # ── 核心方法：prompt ──────────────────────────────────

    async def prompt(
        self,
        goal: str,
        agent_role: Optional[str] = None,
        extra_context: str = "",
        max_turns: Optional[int] = None,
        on_tool_call: Optional[Callable] = None,
    ) -> Dict[str, Any]:
        """
        执行会话：组装提示词并启动 Agent Loop。

        Args:
            goal: 本次执行的目标/任务描述
            agent_role: 指定 Agent 角色（可选，覆盖会话默认角色）
            extra_context: 额外上下文
            max_turns: 最大循环轮次（可选，覆盖配置默认值）
            on_tool_call: 工具调用回调

        Returns:
            {
                "success": bool,
                "content": str,
                "turns": int,
                "error": str | None
            }
        """
        # 更新状态为运行中
        self._update_status("running")

        try:
            # 1. 确定 Agent 配置
            config = self._agent_config
            if agent_role and agent_role != self._record.agent_role:
                config = get_agent_config(agent_role)
                self._record.agent_role = agent_role

            # 2. 组装系统提示词
            prompt_builder = PromptBuilder(config)

            # 3. 仅根会话启用 TaskTool（子 Agent 禁止派发，防止无限嵌套）
            is_root_session = not self._record.parent_id
            if config.mode == "primary" and self._session_manager is not None and is_root_session:
                prompt_builder.enable_task_tool(
                    session_manager=self._session_manager,
                    session_id=self.id,
                    work_dir=self._record.work_dir,
                )

            system_prompt = prompt_builder.build_system_prompt(
                work_dir=self._record.work_dir,
                extra_context=extra_context,
            )

            logger.info(f"[Session] {self.id} 系统提示词: {system_prompt}")
            logger.info(f"[Session] {self.id} 用户输入: {goal}")
            
            # 4.. 获取经过 agent 权限过滤的格式化工具定义
            tool_definitions = prompt_builder.get_filtered_definitions(config)

            # 5. 创建 Agent Loop
            loop = AgentLoop(
                model_client=self._model_client,
                max_turns=max_turns or config.max_turns,
                on_tool_call=on_tool_call,
                tools=prompt_builder.tools,
                tool_definitions=tool_definitions,
            )

            # 6. 执行循环
            result = await loop.run(
                system_prompt=system_prompt,
                user_message=goal,
                messages=self._record.messages if self._record.messages else None,
            )
            
            # 7. 更新消息历史
            if result.get("messages"):
                self._record.messages = result["messages"]
                self._store.update(self._record)

            # 8. 更新状态
            if result.get("success"):
                self._update_status("completed")
            else:
                self._update_status("failed")

            logger.info(
                f"[Session] {self.id} 执行完成，"
                f"轮次={result.get('turns', 0)}，"
                f"成功={result.get('success', False)}"
            )

            return result

        except Exception as e:
            logger.error(f"[Session] {self.id} 执行异常: {e}", exc_info=True)
            self._update_status("failed")
            return {
                "success": False,
                "content": "",
                "turns": 0,
                "error": str(e),
            }

    # ── 内部方法 ──────────────────────────────────────────

    def _update_status(self, status: str):
        """更新会话状态"""
        self._record.status = status
        self._store.update(self._record)


# ============================================================================
#  SessionManager — 会话管理器
# ============================================================================

class SessionManager:
    """
    会话管理器。

    提供会话的创建、删除、查询、列表等管理功能，
    以及 ModelClient 的统一管理。
    """

    def __init__(
        self,
        model_client: ModelClient,
        db_path: Optional[str | Path] = None,
    ):
        """
        Args:
            model_client: LLM 客户端
            db_path: 数据库文件路径
        """
        self._model_client = model_client
        self._store = SessionStore(db_path)
        self._sessions: Dict[str, Session] = {}  # 缓存

        logger.info("[SessionManager] 初始化完成")

    # ── 会话 CRUD ─────────────────────────────────────────

    def create(
        self,
        title: str = "",
        agent_role: str = "build",
        work_dir: str = "",
        parent_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Session:
        """
        创建新会话。

        Args:
            title: 会话标题
            agent_role: Agent 角色
            work_dir: 工作目录
            parent_id: 父会话 ID（subagent 场景）
            metadata: 扩展元信息

        Returns:
            新创建的 Session 实例
        """
        record = SessionRecord(
            title=title,
            agent_role=agent_role,
            work_dir=work_dir,
            parent_id=parent_id,
            metadata=metadata or {},
        )

        # 保存到数据库
        self._store.create(record)

        # 创建 Session 实例
        session = Session(
            record=record,
            store=self._store,
            model_client=self._model_client,
            session_manager=self,
        )

        # 缓存
        self._sessions[record.session_id] = session

        logger.info(
            f"[SessionManager] 创建会话: {record.session_id}，"
            f"角色={agent_role}，标题={title}，"
            f"parent_id={parent_id or '无'}"
        )
        return session

    def get(self, session_id: str) -> Optional[Session]:
        """
        获取会话。

        Args:
            session_id: 会话 ID

        Returns:
            Session 实例，不存在返回 None
        """
        # 先查缓存
        if session_id in self._sessions:
            return self._sessions[session_id]

        # 从数据库加载
        record = self._store.get(session_id)
        if record is None:
            return None

        session = Session(
            record=record,
            store=self._store,
            model_client=self._model_client,
            session_manager=self,
        )
        self._sessions[session_id] = session
        return session

    def delete(self, session_id: str) -> bool:
        """
        删除会话。

        Args:
            session_id: 会话 ID

        Returns:
            是否删除成功
        """
        # 清除缓存
        self._sessions.pop(session_id, None)

        # 从数据库删除
        deleted = self._store.delete(session_id)
        if deleted:
            logger.info(f"[SessionManager] 删除会话: {session_id}")
        return deleted

    def list(
        self,
        parent_id: Optional[str] = None,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Session]:
        """
        列出会话。

        Args:
            parent_id: 父会话 ID 过滤
            status: 状态过滤
            limit: 返回数量限制
            offset: 偏移量

        Returns:
            Session 列表
        """
        records = self._store.list_sessions(
            parent_id=parent_id,
            status=status,
            limit=limit,
            offset=offset,
        )

        sessions = []
        for record in records:
            session = Session(
                record=record,
                store=self._store,
                model_client=self._model_client,
                session_manager=self,
            )
            self._sessions[record.session_id] = session
            sessions.append(session)

        return sessions

    def update_title(self, session_id: str, title: str) -> bool:
        """
        更新会话标题。

        Args:
            session_id: 会话 ID
            title: 新标题

        Returns:
            是否更新成功
        """
        session = self.get(session_id)
        if session is None:
            return False

        session._record.title = title
        return self._store.update(session._record)

    def clear_messages(self, session_id: str) -> bool:
        """
        清空会话消息历史。

        Args:
            session_id: 会话 ID

        Returns:
            是否清空成功
        """
        session = self.get(session_id)
        if session is None:
            return False

        session._record.messages = []
        return self._store.update(session._record)

    # ── 快捷方法 ──────────────────────────────────────────

    async def run_task(
        self,
        goal: str,
        title: str = "",
        agent_role: str = "build",
        work_dir: str = "",
        parent_id: Optional[str] = None,
        max_turns: Optional[int] = None,
    ) -> Dict[str, Any]:
        """
        快捷方法：创建会话并执行任务。

        Args:
            goal: 任务目标
            title: 会话标题（可选）
            agent_role: Agent 角色
            work_dir: 工作目录
            parent_id: 父会话 ID
            max_turns: 最大轮次

        Returns:
            执行结果
        """
        session = self.create(
            title=title or goal[:50],
            agent_role=agent_role,
            work_dir=work_dir,
            parent_id=parent_id,
        )

        return await session.prompt(
            goal=goal,
            max_turns=max_turns,
        )

File: session/session.go
package session

import (
	"context"
	"fmt"
	"log/slog"
	"sync"

	"github.com/user/mini_code/agent"
	"github.com/user/mini_code/provider"
	"github.com/user/mini_code/tools"
	"github.com/user/mini_code/util"
)

type Session struct {
	record           *SessionRecord
	store            *SessionStore
	client           *provider.ModelClient
	agentCfg         *agent.AgentConfig
	manager          *SessionManager
	onToolCallStart  func(name string, params map[string]interface{})
	onToolCallEnd    func(name string, params map[string]interface{}, result string)
	onAssistantReply func(content string, reasoning string)
}

func (s *Session) ID() string             { return s.record.SessionID }
func (s *Session) ParentID() string       { return s.record.ParentID }
func (s *Session) Title() string          { return s.record.Title }
func (s *Session) Status() string         { return s.record.Status }
func (s *Session) WorkDir() string        { return s.record.WorkDir }
func (s *Session) AgentRole() string      { return s.record.AgentRole }
func (s *Session) ChannelKey() string     { return s.record.ChannelKey }
func (s *Session) Record() *SessionRecord { return s.record }

// Prompt runs one agent task and returns the final reply text together with
// the reasoning_content the model produced for it (may be empty).
//
// Panic containment: any panic inside the task (LLM path, compaction, tool
// execution, persistence) is converted into a clean task failure instead of
// crashing the whole process — the server keeps serving other sessions.
func (s *Session) Prompt(ctx context.Context, goal string, maxTurns int) (content string, reasoning string, err error) {
	defer func() {
		if r := recover(); r != nil {
			slog.Error("agent task panicked, contained", "session", s.ID(), "panic", r)
			func() {
				defer func() { recover() }() // never let the bookkeeping itself re-panic
				s.updateStatus("failed")
			}()
			content = ""
			reasoning = ""
			err = fmt.Errorf("agent task crashed (panic: %v)", r)
		}
	}()
	s.updateStatus("running")

	cfg := s.agentCfg
	if cfg == nil {
		var err error
		cfg, err = agent.GetAgentConfig(s.record.AgentRole)
		if err != nil {
			s.updateStatus("failed")
			return "", "", fmt.Errorf("get agent config: %w", err)
		}
	}

	pb := NewPromptBuilder(cfg, s.record.WorkDir)
	isRootSession := s.record.ParentID == ""
	if cfg.Mode == "primary" && s.manager != nil && isRootSession {
		pb.EnableTaskTool(s.manager, s.ID(), s.record.WorkDir)
	}

	systemPrompt := pb.BuildSystemPrompt(s.record.WorkDir, "")
	defs := pb.GetFilteredDefinitions(cfg)

	// Persist the accumulated conversation to the database after every
	// completed turn, so an interrupted task can be resumed from the last
	// good state (interrupt recovery).
	turnHook := func(turn int, messages []provider.Message) {
		if len(messages) == 0 {
			return
		}
		s.record.Messages = messages
		s.store.Update(s.record)
	}

	loop := NewAgentLoop(s.client, maxTurns, pb.Tools, defs, s.onToolCallStart, s.onToolCallEnd, s.onAssistantReply, turnHook)
	if s.manager != nil {
		loop.compactionOn = s.manager.CompactionEnabled()
	}

	slog.Info("session prompt", "id", s.ID(), "goal", goal[:min(80, len(goal))])
	result, err := loop.Run(ctx, systemPrompt, goal, s.record.Messages)
	if err != nil {
		s.updateStatus("failed")
		return "", "", err
	}

	// Always save accumulated messages, even on interrupt.
	if len(result.Messages) > 0 {
		s.record.Messages = result.Messages
		s.store.Update(s.record)
	}

	reasoning = LastAssistantReasoning(result.Messages)

	if result.Interrupted {
		// Mark the session so a later run can resume from the persisted
		// messages (interrupt recovery).
		s.updateStatus("interrupted")
		slog.Info("session interrupted", "id", s.ID(), "turns", result.Turns)
		return result.Content, reasoning, nil
	}

	if result.Success {
		s.updateStatus("completed")
		slog.Info("session done", "id", s.ID(), "turns", result.Turns, "success", true)
		return result.Content, reasoning, nil
	}

	// Propagate the failure reason to the caller so the channel can show it
	// instead of silently stopping.
	s.updateStatus("failed")
	slog.Error("session failed", "id", s.ID(), "turns", result.Turns, "error", result.Error)
	return result.Content, reasoning, fmt.Errorf("agent task failed after %d turns: %s", result.Turns, result.Error)
}

func (s *Session) updateStatus(status string) {
	s.record.Status = status
	s.store.Update(s.record)
}

// CompactSession manually compresses a session's persisted conversation with
// the LLM summarizer, regardless of the auto-compaction threshold. The
// session must not be running. It returns the token count before and after;
// after >= before means nothing was compactable (too short or no safe
// boundary). The compaction is applied to the next LLM call; an in-flight
// task (if one starts while compacting) keeps its own snapshot and the
// manual compaction is simply superseded by it — never data corruption.
func (m *SessionManager) CompactSession(ctx context.Context, sessionID string) (int, int, error) {
	s, err := m.Get(sessionID)
	if err != nil || s == nil {
		return 0, 0, fmt.Errorf("会话不存在: %s", sessionID)
	}
	if s.Status() == "running" {
		return 0, 0, fmt.Errorf("会话正在运行中,请等待任务结束后再压缩")
	}
	msgs := s.Record().Messages
	if len(msgs) == 0 {
		return 0, 0, fmt.Errorf("会话还没有消息,无需压缩")
	}
	before := util.CountMessagesTokens(msgs)

	newMsgs, err := NewContextCompactor(m.clientRef()).Compact(ctx, msgs)
	if err != nil {
		return before, before, fmt.Errorf("压缩失败: %w", err)
	}
	after := util.CountMessagesTokens(newMsgs)
	if after >= before {
		return before, after, nil // nothing compactable
	}
	// Re-check: if a task started while the LLM summary was in flight, do
	// not clobber its working set.
	if s.Status() == "running" {
		return before, after, fmt.Errorf("压缩完成时会话已开始新任务,未应用(下次压缩生效)")
	}
	s.Record().Messages = newMsgs
	m.store.Update(s.Record())
	slog.Info("manual compaction applied", "session", sessionID, "before", before, "after", after)
	return before, after, nil
}

// SessionCallbacks carries the per-session event callbacks. A factory that
// produces one set of callbacks per session ID lets a message hub route each
// session's tool events into the bus, including sub-agent sessions.
type SessionCallbacks struct {
	OnToolStart func(name string, params map[string]interface{})
	OnToolEnd   func(name string, params map[string]interface{}, result string)
	OnReply     func(content string, reasoning string)
}

type SessionManager struct {
	client    *provider.ModelClient
	clientMu  sync.RWMutex
	store     *SessionStore
	sessions  map[string]*Session
	mu        sync.RWMutex
	onToolCallStart  func(name string, params map[string]interface{})
	onToolCallEnd    func(name string, params map[string]interface{}, result string)
	onAssistantReply func(content string, reasoning string)
	callbackFactory  func(sessionID string) *SessionCallbacks

	compactionMu sync.RWMutex
	compactionOn bool // default true (legacy behavior); hot-swappable via SetCompactionEnabled
}

// clientRef returns the current model client.
func (m *SessionManager) clientRef() *provider.ModelClient {
	m.clientMu.RLock()
	defer m.clientMu.RUnlock()
	return m.client
}

// SetClient hot-swaps the model client used by newly created sessions (and
// sub-agent sessions). Sessions already running keep using their existing
// client until they finish.
func (m *SessionManager) SetClient(c *provider.ModelClient) {
	m.clientMu.Lock()
	m.client = c
	m.clientMu.Unlock()
}

// SetCompactionEnabled hot-toggles automatic context compression. It applies
// to turns started after the call; in-flight turns keep their setting.
func (m *SessionManager) SetCompactionEnabled(on bool) {
	m.compactionMu.Lock()
	m.compactionOn = on
	m.compactionMu.Unlock()
}

// CompactionEnabled reports the current compaction toggle (default true).
func (m *SessionManager) CompactionEnabled() bool {
	m.compactionMu.RLock()
	defer m.compactionMu.RUnlock()
	return m.compactionOn
}

func NewSessionManager(client *provider.ModelClient, dbPath string) (*SessionManager, error) {
	store, err := NewSessionStore(dbPath)
	if err != nil {
		return nil, err
	}
	return &SessionManager{
		client:       client,
		store:        store,
		sessions:     make(map[string]*Session),
		compactionOn: true,
	}, nil
}

// NewDefaultSessionManager creates a SessionManager using the default database path (~/.mini_code/agent.db)
func NewDefaultSessionManager(client *provider.ModelClient) (*SessionManager, error) {
	store, err := NewDefaultSessionStore()
	if err != nil {
		return nil, err
	}
	return &SessionManager{
		client:       client,
		store:        store,
		sessions:     make(map[string]*Session),
		compactionOn: true,
	}, nil
}

// SetToolCallCallbacks sets global callbacks for tool call events (legacy,
// applies to every session).
func (m *SessionManager) SetToolCallCallbacks(onStart func(string, map[string]interface{}), onEnd func(string, map[string]interface{}, string), onReply func(string, string)) {
	m.onToolCallStart = onStart
	m.onToolCallEnd = onEnd
	m.onAssistantReply = onReply
}

// SetSessionCallbackFactory registers a per-session callback factory. When
// set, it takes precedence over the global callbacks: each session (including
// sub-agent sessions) gets its own callbacks resolved by session ID.
func (m *SessionManager) SetSessionCallbackFactory(f func(sessionID string) *SessionCallbacks) {
	m.mu.Lock()
	m.callbackFactory = f
	m.mu.Unlock()
}

// callbacksFor resolves the effective callbacks for a session ID.
func (m *SessionManager) callbacksFor(sessionID string) *SessionCallbacks {
	m.mu.RLock()
	f := m.callbackFactory
	legacyStart, legacyEnd, legacyReply := m.onToolCallStart, m.onToolCallEnd, m.onAssistantReply
	m.mu.RUnlock()

	if f != nil {
		if cbs := f(sessionID); cbs != nil {
			return cbs
		}
	}
	if legacyStart != nil || legacyEnd != nil || legacyReply != nil {
		return &SessionCallbacks{OnToolStart: legacyStart, OnToolEnd: legacyEnd, OnReply: legacyReply}
	}
	return nil
}

// newSession builds a Session with its event callbacks resolved.
func (m *SessionManager) newSession(record *SessionRecord) *Session {
	cfg, _ := agent.GetAgentConfig(record.AgentRole)
	cbs := m.callbacksFor(record.SessionID)
	s := &Session{
		record:   record,
		store:    m.store,
		client:   m.clientRef(),
		agentCfg: cfg,
		manager:  m,
	}
	if cbs != nil {
		s.onToolCallStart = cbs.OnToolStart
		s.onToolCallEnd = cbs.OnToolEnd
		s.onAssistantReply = cbs.OnReply
	}
	return s
}

func (m *SessionManager) CreateSession(title, agentRole, workDir string, parentID *string) (tools.SessionPrompter, error) {
	return m.CreateSessionRecord(title, agentRole, workDir, parentID)
}

// CreateSessionRecord creates a session and returns the concrete *Session
// (use this when the caller needs session-level APIs such as ID()).
func (m *SessionManager) CreateSessionRecord(title, agentRole, workDir string, parentID *string) (*Session, error) {
	pid := ""
	if parentID != nil {
		pid = *parentID
	}
	record := NewSessionRecord(title, agentRole, workDir, pid)
	if err := m.store.Create(record); err != nil {
		return nil, err
	}

	session := m.newSession(record)

	m.mu.Lock()
	m.sessions[record.SessionID] = session
	m.mu.Unlock()

	slog.Info("session created", "id", record.SessionID, "role", agentRole, "parent", pid)
	return session, nil
}

func (m *SessionManager) Get(sessionID string) (*Session, error) {
	m.mu.RLock()
	if s, ok := m.sessions[sessionID]; ok {
		m.mu.RUnlock()
		return s, nil
	}
	m.mu.RUnlock()

	record, err := m.store.Get(sessionID)
	if err != nil || record == nil {
		return nil, err
	}

	session := m.newSession(record)

	m.mu.Lock()
	m.sessions[sessionID] = session
	m.mu.Unlock()
	return session, nil
}

// GetByChannelKey returns the most recently updated session bound to the given
// channel key (e.g. a DingTalk conversation id). It returns nil when the key
// is not bound to any session.
func (m *SessionManager) GetByChannelKey(key string) (*Session, error) {
	if key == "" {
		return nil, nil
	}

	record, err := m.store.GetByChannelKey(key)
	if err != nil || record == nil {
		return nil, err
	}
	session := m.newSession(record)
	m.mu.Lock()
	m.sessions[record.SessionID] = session
	m.mu.Unlock()
	return session, nil
}

// SetChannelKey binds a stable external channel key (e.g. a DingTalk
// conversation id) to a session, so future messages carrying that key are
// routed to the same session.
func (m *SessionManager) SetChannelKey(sessionID, key string) error {
	if err := m.store.SetChannelKey(sessionID, key); err != nil {
		return err
	}
	// Keep the cached in-memory record in sync so that a later whole-row
	// persistence (store.Update) cannot clobber the binding with a stale
	// empty key.
	m.mu.Lock()
	if s, ok := m.sessions[sessionID]; ok {
		s.record.ChannelKey = key
	}
	m.mu.Unlock()
	return nil
}

func (m *SessionManager) Delete(sessionID string) error {
	m.mu.Lock()
	delete(m.sessions, sessionID)
	m.mu.Unlock()
	return m.store.Delete(sessionID)
}

func (m *SessionManager) List(parentID, status string, limit, offset int) ([]*Session, error) {
	records, err := m.store.ListSessions(parentID, status, limit, offset)
	if err != nil {
		return nil, err
	}
	var sessions []*Session
	for _, r := range records {
		sessions = append(sessions, m.newSession(r))
	}
	return sessions, nil
}

func (m *SessionManager) Close() error {
	return m.store.Close()
}

// GetStore returns the session store for external access
func (m *SessionManager) GetStore() *SessionStore {
	return m.store
}
